src/attachment.rs

```diff
@@ -0,0 +1,347 @@
+pub extern "C" fn z_bytes_map_insert_by_copy(
+    this: &z_owned_bytes_map_t,
+    key: z_bytes_t,
+    value: z_bytes_t,
+) {
+    let this = unsafe { &mut *this.get() };
+    if let (Some(this), Some(key), Some(value)) = (this.as_mut(), key.as_slice(), value.as_slice())
+    {
+        this.insert(Cow::Owned(key.to_owned()), Cow::Owned(value.to_owned()));
+    }
+}
+
+/// Associates `value` to `key` in the map, aliasing them.
+///
+/// Note that once `key` is aliased, reinserting at the same key may alias the previous instance, or the new instance of `key`.
+///
+/// Calling this with `NULL` or the gravestone value is undefined behaviour.
+#[no_mangle]
+pub extern "C" fn z_bytes_map_insert_by_alias(
+    this: &z_owned_bytes_map_t,
+    key: z_bytes_t,
+    value: z_bytes_t,
+) {
+    let this = unsafe { &mut *this.get() };
+    if let (Some(this), Some(key), Some(value)) = (this.as_mut(), key.as_slice(), value.as_slice())
+    {
+        unsafe {
+            this.insert(
+                Cow::Borrowed(core::mem::transmute(key)),
+                Cow::Borrowed(core::mem::transmute(value)),
+            )
+        };
+    }
+}
+
+/// Iterates over the key-value pairs in the map.
+///
+/// `body` will be called once per pair, with `ctx` as its last argument.
+/// If `body` returns a non-zero value, the iteration will stop immediately and the value will be returned.
+/// Otherwise, this will return 0 once all pairs have been visited.
+/// `body` is not given ownership of the key nor value, which alias the pairs in the map.
+/// It is safe to keep these aliases until existing keys are modified/removed, or the map is destroyed.
+/// Note that this map is unordered.
+///
+/// Calling this with `NULL` or the gravestone value is undefined behaviour.
+#[no_mangle]
+pub extern "C" fn z_bytes_map_iter(
+    this: &z_owned_bytes_map_t,
+    body: z_attachment_iter_body_t,
+    ctx: *mut c_void,
+) -> i8 {
+    let this = unsafe { &*this.get() };
+    if let Some(this) = this.as_ref() {
+        for (key, value) in this.iter() {
+            let result = body(key.as_ref().into(), value.as_ref().into(), ctx);
+            if result != 0 {
+                return result;
+            }
+        }
+    }
+    0
+}
+
+const Z_BYTES_MAP_ITERATION_DRIVER: z_attachment_iter_driver_t =
+    Some(unsafe { core::mem::transmute(z_bytes_map_iter as extern "C" fn(_, _, _) -> i8) });
+
+//pub(crate) extern "C" fn Z_BYTES_MAP_ITERATION_DRIVER(
+//    this: *const c_void,
+//    body: z_attachment_iter_body_t,
+//    ctx: *mut c_void,
+//) -> i8 {
+//    let r: &z_owned_bytes_map_t = unsafe { &*(this as *mut z_owned_bytes_map_t) };
+//    z_bytes_map_iter(r, body, ctx)
+//}
+
+pub(crate) extern "C" fn insert_in_attachment(
+    key: z_bytes_t,
+    value: z_bytes_t,
+    ctx: *mut c_void,
+) -> i8 {
+    let attachments_ref: &mut Attachment = unsafe { &mut *(ctx as *mut Attachment) };
+    attachments_ref.insert(key.as_slice().unwrap(), value.as_slice().unwrap());
```

Comment:
I've just added `AttachmentBuilder` to the Rust API to make inserts _much_ faster, could you change this (and callers) to use that instead of `Attachment`? `Attachment::insert` is awful slow, but that can't really be helped much yet.